### ISSUE TITLE
Crash with Encrypted DB & Shared Container

### ISSUE THREAD

**Issue by jayahariv** [couchbase/couchbase-lite-core#647]
Steps: 
1. App with DB in the shared container, also it should be encrypted. 
2. When the app is send to background, the app crashes, with console message, 
```
[CBLiteTaskExten:1441] Forcing crash report with description: CBLiteTaskExten:1441 was suspended with locked system files:
/private/var/mobile/Containers/Shared/AppGroup/FCDB5594-893D-4832-B1C9-755519574284/CBLite/demo.cblite2/db.sqlite3
```
- Only happens in device

Easily replicated with adding an encryptionKey to this [extension project](https://blog.couchbase.com/sharing-data-ios-app-extensions-sync-capability/) and run on device. Will get the log from device console log. 

[SQLCipherVsSharedData](https://github.com/signalapp/SQLCipherVsSharedData) explained in detail how this issue happens with SQLCipher.

SQLCipher team has analyzed and shared what really happens, and how iOS can't identify it as a sqlite file. [comment from Stephen](https://github.com/sqlcipher/sqlcipher/issues/255#issuecomment-355063368)

Link: [#2241](https://github.com/couchbase/couchbase-lite-ios/issues/2241)

**Comment by snej** [comment]
The workarounds aren't very appealing. Turning off WAL mode would hurt performance, and reduce availability by blocking reads during a write. Hacking the SQLite Encryption Extension to leave the first few bytes of the file alone is kind of scary.

Honestly, I'd advise (as I have before) to turn off DB encryption. All iOS app files are already encrypted.

**Comment by snej** [comment]
What about closing the database when the app is backgrounded? That seems simple enough.

**Comment by jayahariv** [comment]
yes, closing the database will avoid the crash. thats a valid workaround.

**Comment by timritzer** [comment]
For what it's worth, our use case definitely doesn't let us turn off encryption on the DB. It loses the FIPS compliance required for our healthcare based application. 

We thankfully can wait to use it from an app group, but it would be really helpful for future architecture.

**Comment by snej** [comment]
@jayahariv There's nothing wrong with SQLite.

Apple's description of the exception is:

>The exception code 0xdead10cc indicates that an application has been terminated by the OS because it held on to a file lock or sqlite database lock during suspension. If your application is performing operations on a locked file or sqlite database at suspension time, it must request additional background execution time to complete those operations and relinquish the lock before suspending.

So the answer is that when the app is notified it's being suspended, it needs to close the database, or request extra time to finish a task like replication. We can do this in CBL by listening for the appropriate notification.

**Comment by snej** [comment]
@timritzer All iOS apps' files are already encrypted at rest with Apple's FIPS-compliant AES codec. Do you have explicit confirmation that this isn't sufficient for your app's FIPS compliance? (I haven't been able to find something that says conclusively one way or the other.)

**Comment by timritzer** [comment]
Sorry I misspoke. Not FIPS. AES, and Apple's hardware (and kernel) crypto modules for AES are both FIPS approved. And if you just need FIPS 140 compliance you are golden. 

It's the NIST 800-63 Assurance level required for offline Health information storage. Because we are storing Medical PHI on device, we are required to have NIST LOA of 3, which requires a few things relating to key management that we don't get by the apple provided key. 

I would provide links, but apparently the NIST web archives are down until our government is no longer shut down. It's a really boring read with lots of NIST jargon anyways.

The end result is that we need to control the key used to protect the data, and relying on Apple's encryption does not give us that. So it isn't sufficient to meet the requirements for offline PHI access.

As mentioned before though, we don't need immediate use of the stores in shared data, so it isn't a huge deal to us right this instant. However as we move towards our ultimate desired architecture, I think it would really help to be able to have app groups share the data.

**Comment by snej** [comment]
@timritzer Gotcha. I'm aware that healthcare apps tend to need extra encryption.

I'm going to move this issue to iOS because it's not a LiteCore-level problem -- the workaround will need to go into the iOS codebase, to detect the backgrounding notification and close the DB.
